File: Chapter 03/Python/section_3_3_1_srp/after/tactics_engine.py
from .player import Player

"""
Responsibility 2: Tactical / AI Logic.
* ARCHITECTURE NOTE: This class is a "Service." By moving tactical decisions 
here, we've decoupled the 'Intelligence' of the game from the 'Actor' (the Player). 
* In Python, this allows you to easily swap out this engine for a different 
version—perhaps one driven by Machine Learning or a simple rule-based system—
without ever needing to modify the Player class itself.
"""
class TacticsEngine:
    def determine_best_position(self, player: 'Player'):
        """
        Analyzes the game state to determine the optimal position for a specific player.
        """
        print(f"  [Tactics] Calculating optimal court position for {player.name}...")

File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_1_srp/after/TacticsEngine.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_1_srp.after;

/**
 * Responsibility 2: Tactical / AI Logic.
 * * ARCHITECTURE NOTE: Separating tactics into its own engine means we can 
 * easily swap out algorithms (e.g., switching from "Aggressive" to "Defensive" 
 * tactics) without ever touching the Player class. It also makes this complex 
 * logic incredibly easy to unit test in isolation.
 */
public class TacticsEngine {
    
    /**
     * Analyzes the game state to determine the optimal position for a specific player.
     * * @param player The player being analyzed.
     */
    public void determineBestPosition(Player player) {
        System.out.println("  [Tactics] Calculating optimal court position for " + player.name + "...");
    }
}